src/scanner.c

```diff
@@ -231,73 +214,141 @@ ts_scanner_start_scan(ScannerCtx *ctx, InternalScannerCtx *ictx)
 		 * mode.
 		 */
 		ctx->snapshot = RegisterSnapshot(GetSnapshotData(SnapshotSelf));
-		ictx->registered_snapshot = true;
+		ctx->internal.registered_snapshot = true;
+	}
+}
+
+TSDLLEXPORT Relation
+ts_scanner_open(ScannerCtx *ctx)
+{
+	Scanner *scanner = scanner_ctx_get_scanner(ctx);
+
+	Assert(NULL == ctx->tablerel);
+	prepare_scan(ctx);
+
+	return scanner->openscan(ctx);
+}
+
+/*
+ * Start either a heap or index scan depending on the information in the
+ * ScannerCtx. ScannerCtx must be setup by caller with the proper information
+ * for the scan, including filters and callbacks for found tuples.
+ */
+TSDLLEXPORT void
+ts_scanner_start_scan(ScannerCtx *ctx)
+{
+	InternalScannerCtx *ictx = &ctx->internal;
+	Scanner *scanner;
+	TupleDesc tuple_desc;
+
+	if (ictx->started)
+	{
+		Assert(!ictx->ended);
+		Assert(ctx->tablerel);
+		Assert(OidIsValid(ctx->table));
+		return;
+	}
```

Comment:
I see lots of uncovered lines here and there, are we missing some tests?

Reply:
This one I might be able to do something about, but most others are either:
* Freeing materialized tuples in scans (this seldom happens and is not up to us):
```
if (should_free)
						heap_freetuple(tuple);
```
* Or, a chunk was concurrently removed before got a lock during scans. We could try to add an isolation test for that case, but it is quite difficult to test for.

Reply:
Added a new isolation test and scanner native test to improve the coverage.